File: app/src/main/java/me/magnum/melonds/parcelables/RomParcelable.java
package me.magnum.melonds.parcelables;

import android.os.Parcel;
import android.os.Parcelable;
import me.magnum.melonds.model.Rom;
import me.magnum.melonds.model.RomConfig;

public class RomParcelable implements Parcelable {
    private Rom rom;

    public RomParcelable(Rom rom) {
        this.rom = rom;
    }

    protected RomParcelable(Parcel in) {
        RomConfig romConfig = new RomConfig();
        this.rom = new Rom(in.readString(), in.readString(), romConfig);
        romConfig.setLoadGbaCart(in.readInt() == 1);
        romConfig.setGbaCartPath(in.readString());
        romConfig.setGbaSavePath(in.readString());
    }

    public Rom getRom() {
        return rom;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(rom.getName());
        dest.writeString(rom.getPath());
        dest.writeInt(rom.getConfig().loadGbaCart() ? 1 : 0);
        dest.writeString(rom.getConfig().getGbaCartPath());
        dest.writeString(rom.getConfig().getGbaSavePath());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<RomParcelable> CREATOR = new Creator<RomParcelable>() {
        @Override
        public RomParcelable createFromParcel(Parcel in) {
            return new RomParcelable(in);
        }

        @Override
        public RomParcelable[] newArray(int size) {
            return new RomParcelable[size];
        }
    };
}

File: app/src/main/java/me/magnum/melonds/parcelables/RomParcelable.kt
package me.magnum.melonds.parcelables

import android.os.Parcel
import android.os.Parcelable
import androidx.core.net.toUri
import me.magnum.melonds.domain.model.rom.Rom
import me.magnum.melonds.extensions.parcelable
import java.util.*
import kotlin.time.Duration.Companion.milliseconds

class RomParcelable : Parcelable {
    var rom: Rom
        private set

    constructor(rom: Rom) {
        this.rom = rom
    }

    private constructor(parcel: Parcel) {
        val name = parcel.readString()
        val developerName = parcel.readString()!!
        val fileName = parcel.readString()
        val uri = parcel.readString()!!.toUri()
        val parentTreeUri = parcel.readString()?.toUri()
        val lastPlayed = parcel.readLong().let { if (it == (-1).toLong()) null else Date(it) }
        val romConfig = parcel.parcelable<RomConfigParcelable>()
        val isDsiWareTitle = parcel.readInt() == 1
        val retroAchievementsHash = parcel.readString()!!
        val totalPlayTime = parcel.readLong().milliseconds
        rom = Rom(name!!, developerName, fileName!!, uri, parentTreeUri, romConfig!!.romConfig, lastPlayed, isDsiWareTitle, retroAchievementsHash, totalPlayTime)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(rom.name)
        dest.writeString(rom.developerName)
        dest.writeString(rom.fileName)
        dest.writeString(rom.uri.toString())
        dest.writeString(rom.parentTreeUri?.toString())
        dest.writeLong(rom.lastPlayed?.time ?: -1)
        dest.writeParcelable(RomConfigParcelable(rom.config), 0)
        dest.writeInt(if (rom.isDsiWareTitle) 1 else 0)
        dest.writeString(rom.retroAchievementsHash)
        dest.writeLong(rom.totalPlayTime.inWholeMilliseconds)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<RomParcelable> {
        override fun createFromParcel(parcel: Parcel): RomParcelable {
            return RomParcelable(parcel)
        }

        override fun newArray(size: Int): Array<RomParcelable?> {
            return arrayOfNulls(size)
        }
    }
}